journal/src/main/java/io/zeebe/journal/file/SegmentedJournal.java

```diff
@@ -0,0 +1,550 @@
+import java.util.concurrent.ConcurrentSkipListMap;
+import org.agrona.DirectBuffer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A file based journal. The journal is split into multiple segments files. */
+public class SegmentedJournal implements Journal {
+  private static final int SEGMENT_BUFFER_FACTOR = 3;
+  private final JournalMetrics journalMetrics;
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  private final String name;
+  private final File directory;
+  private final int maxSegmentSize;
+  private final int maxEntrySize;
+  private final NavigableMap<Long, JournalSegment> segments = new ConcurrentSkipListMap<>();
+  private final Collection<SegmentedJournalReader> readers = Sets.newConcurrentHashSet();
+  private volatile JournalSegment currentSegment;
+  private volatile boolean open = true;
+  private final long minFreeDiskSpace;
+  private final JournalIndex journalIndex;
+  private final SegmentedJournalWriter writer;
+
+  public SegmentedJournal(
+      final String name,
+      final File directory,
+      final int maxSegmentSize,
+      final int maxEntrySize,
+      final long minFreeSpace,
+      final JournalIndex journalIndex) {
+    this.name = checkNotNull(name, "name cannot be null");
+    this.directory = checkNotNull(directory, "directory cannot be null");
+    this.maxSegmentSize = maxSegmentSize;
+    this.maxEntrySize = maxEntrySize;
+    journalMetrics = new JournalMetrics(name);
+    minFreeDiskSpace = minFreeSpace;
+    this.journalIndex = journalIndex;
+    open();
+    writer = new SegmentedJournalWriter(this);
+  }
+
+  /**
+   * Returns a new SegmentedJournal builder.
+   *
+   * @return A new Segmented journal builder.
+   */
+  public static SegmentedJournalBuilder builder() {
+    return new SegmentedJournalBuilder();
+  }
+
+  @Override
+  public JournalRecord append(final long asqn, final DirectBuffer data) {
+    return writer.append(asqn, data);
+  }
+
+  @Override
+  public void append(final JournalRecord record) {
+    writer.append(record);
+  }
+
+  @Override
+  public void deleteAfter(final long indexExclusive) {
+    writer.deleteAfter(indexExclusive);
+  }
+
+  @Override
+  public void deleteUntil(final long index) {
+    final Map.Entry<Long, JournalSegment> segmentEntry = segments.floorEntry(index);
+    if (segmentEntry != null) {
+      final SortedMap<Long, JournalSegment> compactSegments =
+          segments.headMap(segmentEntry.getValue().index());
+      if (!compactSegments.isEmpty()) {
+        log.debug("{} - Compacting {} segment(s)", name, compactSegments.size());
+        for (final JournalSegment segment : compactSegments.values()) {
+          log.trace("Deleting segment: {}", segment);
+          segment.compactIndex(index);
+          segment.close();
+          segment.delete();
+          journalMetrics.decSegmentCount();
+        }
+        compactSegments.clear();
+      }
+    }
+    resetHead(getFirstSegment().index());
```

Comment:
You move this out of the conditional branch is this intended?

Reply:
No. This was by mistake. Fixed it. :+1: